proptest-derive/src/error.rs

```diff
@@ -0,0 +1,512 @@
+    /// Ensures that an explicit strategy or value is not present on a unit variant.
+    use attr::StratMode::*;
+    match attrs.strategy {
+        Arbitrary   => {},
+        Strategy(_) => strategy_on_unit_variant(ctx, "strategy"),
+        Value(_)    => strategy_on_unit_variant(ctx, "value"),
+    }
+
+    if attrs.params.is_set() {
+        params_on_unit_variant(ctx)
+    }
+
+    if !attrs.filter.is_empty() {
+        filter_on_unit_variant(ctx)
+    }
+}
+
+/// Ensures that parameters or filter is not present on a unit struct.
+pub fn if_present_on_unit_struct(ctx: Ctx, attrs: &ParsedAttributes) {
+    if attrs.params.is_set() {
+        params_on_unit_struct(ctx)
+    }
+
+    if !attrs.filter.is_empty() {
+        filter_on_unit_struct(ctx)
+    }
+}
+
+/// Ensures that skip is not present on `item`.
+pub fn if_skip_present(ctx: Ctx, attrs: &ParsedAttributes, item: &str) {
+    if attrs.skip {
+        illegal_skip(ctx, item)
+    }
+}
+
+/// Ensures that a weight is not present on `item`.
+pub fn if_weight_present(ctx: Ctx, attrs: &ParsedAttributes, item: &str) {
+    if attrs.weight.is_some() {
+        illegal_weight(ctx, item)
+    }
+}
+
+//==============================================================================
+// Messages
+//==============================================================================
+
+use std::fmt::Display;
+
+#[derive(Debug)]
+pub struct Fatal;
+pub type DeriveResult<T> = Result<T, Fatal>;
+pub type Ctx<'ctx> = &'ctx mut Context;
+
+pub struct Context {
+    errors: Vec<String>,
+}
+
+impl Context {
+    pub fn new() -> Self {
+        Self {
+            errors: Vec::new(),
+        }
+    }
+
+    pub fn error<T: Display>(&mut self, msg: T) {
+        self.errors.push(msg.to_string());
+    }
+
+    pub fn fatal<T: Display, A>(&mut self, msg: T) -> DeriveResult<A> {
+        self.error(msg);
+        Err(Fatal)
+    }
+
+    pub fn check(mut self) -> Result<(), TokenStream> {
+        fn compile_error(msg: &str) -> TokenStream {
+            quote! {
+                compile_error!(#msg);
+            }
+        }
+
+        match self.errors.len() {
+            0 => Ok(()),
+            1 => Err(compile_error(&self.errors.pop().unwrap())),
+            n => {
+                let mut msg = format!("{} errors:", n);
+                for err in self.errors {
+                    msg.push_str("\n\t# ");
+                    msg.push_str(&err);
+                }
+                Err(compile_error(&msg))
+            }
+        }
+    }
+}
+
+//==============================================================================
+// Messages
+//==============================================================================
+
+macro_rules! mk_err_msg {
+    ($code: ident, $msg: expr) => {
+        concat!(
+            "[proptest_derive, ", stringify!($code), "]",
+            " during #[derive(Arbitrary)]:\n",
+            $msg,
+            " Please see: https://PATH/TO/foo#", stringify!($code),
```

Comment:
We'll need to update this once the location of the docs are selected. (Just making a note so this is easier to see later.)

Reply:
Nicely spotted; I might have forgotten about this otherwise. =P